CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/GalleryFragment.kt

```diff
@@ -52,29 +49,28 @@ class GalleryFragment internal constructor() : Fragment() {
     /** AndroidX navigation arguments */
     private val args: GalleryFragmentArgs by navArgs()
 
-    private lateinit var mediaList: MutableList<File>
+    private lateinit var mediaList: MutableList<MediaStoreFile>
 
     /** Adapter class used to present a fragment containing one photo or video as a page */
-    inner class MediaPagerAdapter(fm: FragmentManager) : FragmentStatePagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
-        override fun getCount(): Int = mediaList.size
-        override fun getItem(position: Int): Fragment = PhotoFragment.create(mediaList[position])
-        override fun getItemPosition(obj: Any): Int = POSITION_NONE
+    inner class MediaPagerAdapter(fm: FragmentManager) : FragmentStateAdapter(fm, lifecycle) {
+        override fun getItemCount(): Int = mediaList.size
+        override fun createFragment(position: Int): Fragment = PhotoFragment.create(mediaList[position])
+        override fun getItemId(position: Int): Long {
+            return mediaList[position].id
+        }
+        override fun containsItem(itemId: Long): Boolean {
+            return null != mediaList.firstOrNull { it.id == itemId }
+        }
     }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
-        // Mark this as a retain fragment, so the lifecycle does not get restarted on config change
-        retainInstance = true
-
-        // Get root directory of media from navigation arguments
-        val rootDirectory = File(args.rootDirectory)
-
-        // Walk through all files in the root directory
-        // We reverse the order of the list to present the last photos first
-        mediaList = rootDirectory.listFiles { file ->
-            EXTENSION_WHITELIST.contains(file.extension.toUpperCase(Locale.ROOT))
-        }?.sortedDescending()?.toMutableList() ?: mutableListOf()
+        runBlocking {
```

Comment:
we should avoid `runBlocking` because it's going to block the main thread
maybe
```
private var mediaList: MutableList<MediaStoreFile> = emptyList()
...
lifecycleScope.launch {
    mediaList = MediaStoreUtils(requireContext()).getImages()
    adapter.mediaItems = mediaList
}
```

Reply:
I originally tried exactly the code you have, but the app crashed in `onViewCreated` when `mediaList` wasn't defined yet. I've tested the app with runBlocking and there is no lag, even with over 100 photos to show in the gallery carousel. The only other option I can think of is async/await, but that would put most of the code in `onViewCreated` in a launch block after the call to `await` anyway. So that seems to have latency issues as well. Still, perhaps that's better because the main thread isn't completely blocked, even though nothing can be shown on the photo gallery until the photos have been loaded. I can implement it that way if you think it's preferable. LMK.

Reply:
I implemented this with `async` in `onCreate` and the app is working as expected:

```
mediaListDeferred = lifecycleScope.async {
    // Get images this app has access to from MediaStore
    val mediaStoreUtils = MediaStoreUtils(requireContext())
    mediaStoreUtils.getImages()
}
```
Then I put most of the `onViewCreated` code after `mediaList = mediaListDeferred.await()` in a `launch` block. I can add that commit if this approach sounds good to you.

Reply:
Thanks for the offline conversation on this. I've implemented your suggested changes and removed `runBlocking` completely.